numpy/_core/src/multiarray/iterators.c

```diff
@@ -561,195 +489,157 @@ iter_subscript_int(PyArrayIterObject *self, PyArrayObject *ind,
-                                     dtype,
-                                     1, &ii,
-                                     NULL, NULL, 0,
-                                     (PyObject *)self->ao);
-            return (PyObject *)ret;
+            ret = PyArray_NewFromDescr(Py_TYPE(self->ao),
+                                       dtype,
+                                       1, &ii,
+                                       NULL, NULL, 0,
+                                       (PyObject *)self->ao);
+            goto finish;
         }
     }
 
-    dtype = PyArray_DESCR(self->ao);
-    size = dtype->elsize;
+    PyArray_ITER_RESET(self);
 
-    /* set up a cast to handle item copying */
+    if (index_type == HAS_INTEGER) {
+        if (check_and_adjust_index(&indices[0].value, self->size, -1, NULL) < 0) {
+            goto finish;
+        }
+        PyArray_ITER_GOTO1D(self, indices[0].value);
+        ret = PyArray_ToScalar(self->dataptr, self->ao);
+        PyArray_ITER_RESET(self);
+        goto finish;
+    }
 
+    /* set up a cast to handle item copying */
     NPY_ARRAYMETHOD_FLAGS transfer_flags = 0;
     npy_intp one = 1;
+
     /* We can assume the newly allocated output array is aligned */
     int is_aligned = IsUintAligned(self->ao);
     if (PyArray_GetDTypeTransferFunction(
-                is_aligned, size, size, dtype, dtype, 0, &cast_info,
+                is_aligned, dtype_size, dtype_size, dtype, dtype, 0, &cast_info,
                 &transfer_flags) < 0) {
-        goto fail;
+        goto finish;
     }
 
-    /* Check for Integer or Slice */
-    if (PyLong_Check(ind) || PySlice_Check(ind)) {
-        start = parse_index_entry(ind, &step_size, &n_steps,
-                                  self->size, 0, 1);
-        if (start == -1) {
-            goto fail;
-        }
-        if (n_steps == ELLIPSIS_INDEX || n_steps == NEWAXIS_INDEX) {
-            PyErr_SetString(PyExc_IndexError,
-                            "cannot use Ellipsis or newaxes here");
-            goto fail;
+    if (index_type & HAS_SLICE) {
+        if (PySlice_GetIndicesEx(indices[0].object,
+                                 self->size,
+                                 &start, &stop, &step, &n_steps) < 0) {
+            goto finish;
         }
+
         PyArray_ITER_GOTO1D(self, start);
-        if (n_steps == SINGLE_INDEX) { /* Integer */
-            PyObject *tmp;
-            tmp = PyArray_ToScalar(self->dataptr, self->ao);
-            PyArray_ITER_RESET(self);
-            NPY_cast_info_xfree(&cast_info);
-            return tmp;
-        }
         Py_INCREF(dtype);
-        ret = (PyArrayObject *)PyArray_NewFromDescr(Py_TYPE(self->ao),
+        ret = PyArray_NewFromDescr(Py_TYPE(self->ao),
                                  dtype,
                                  1, &n_steps,
                                  NULL, NULL,
                                  0, (PyObject *)self->ao);
```

Comment:
Not your problem, so ignore.  I think there might be a bit of fuzziness around the fact that sometimes `dtypes` get's replaced on array creation.